### PR TITLE
#180: do not allow dupe forms in the IBDO stack

### PR DESCRIPTION
**Pull request by andrewbp**
This work simply adds a "remove(newForm)" call before any attempt to add a new form to the IBDO's form stack. This will remove any already existing matching forms (or do nothing if it doesn't already exist) and then continue as before. The test I wrote adds the same form multiple times in different ways and then checks to see that only one of that form exists in the stack.

As an aside, it may be worth revisiting BaseDataObject for a little bit of refactoring, but that's outside the scope of this ticket.

### CODE REVIEW

**Comment by andrewbp** [comment]
In RoutingAlgorithmTest, testErrorHandlingErrorPopsAllCurrentForms() is currently failing after this change because it is expecting to see at least two ERRORs in the form stack, which is now impossible. There are a few ways to fix this:

1) Allow for multiple ERRORs to exist in the stack
2) Change the test conditions/answers
3) Change the check in MobileAgent, line 454

For now, I will implement allowing multiple ERRORs in the stack as it is the easiest to implement and still keep with current functionality.

File: src/main/java/emissary/core/BaseDataObject.java [hunk 1]
```diff
@@ -440,10 +441,17 @@ public int enqueueCurrentForm(final String newForm) {
             throw new IllegalArgumentException("caller attempted to enqueue a null form value");
         }
 
+        checkForDuplicates(newForm);
```
Comment (dev-mlb) [inline src/main/java/emissary/core/BaseDataObject.java:444]: Thinking out loud here, is it better to remove the form or not add it? Essentially, you could be kicking a form to the back of the line - I don't know if that is good or bad. Possibly unintended consequences? Something to think about.

Comment (andrewbp) [reply src/main/java/emissary/core/BaseDataObject.java:444]: The intention was to push the form to the top if for whatever reason it was encountered again, but I may have gone about it incorrectly.

**Comment by jpdahlke** [comment]
After looking over the current code and talking it over with @dev-mlb a bit, I think if we wanted to do anything now we should start enumerating these instances with logs so we can see how often it happens and under which situations before we take actions to prevent. The ticket does ask for logging updates and confirmation and that might be a better first step.

I'm concerned that there are natural situations we need to know about and then there is the noted ERROR form handling scenario. It may be the case that we are safe from running into multiple ERROR forms with changes we've made to routing and removing moveTo, but the kind of code I would have replaced https://github.com/NationalSecurityAgency/emissary/blob/7.7.0/src/main/java/emissary/core/MobileAgent.java#L454 with was actually bug fixed to be the current state which makes me wonder what I'm missing.

**Comment by jpdahlke** [comment]
> > After looking over the current code and talking it over with @dev-mlb a bit, I think if we wanted to do anything now we should start enumerating these instances with logs so we can see how often it happens and under which situations before we take actions to prevent. The ticket does ask for logging updates and confirmation and that might be a better first step.
> > I'm concerned that there are natural situations we need to know about and then there is the noted ERROR form handling scenario. It may be the case that we are safe from running into multiple ERROR forms with changes we've made to routing and removing moveTo, but the kind of code I would have replaced https://github.com/NationalSecurityAgency/emissary/blob/7.7.0/src/main/java/emissary/core/MobileAgent.java#L454 with was actually bug fixed to be the current state which makes me wonder what I'm missing.
> 
> Just to make sure I'm interpreting correctly, we should scratch this pull request and instead just log instances of duplicate forms?

Yes, let's go that route for now to better understand this behavior.

**Comment by andrewbp** [comment]
Closing this and opening a new pull request: https://github.com/NationalSecurityAgency/emissary/pull/304

**Pull request closed without merging**